[SPARK-10905][SparkR]: Export freqItems() for DataFrameStatFunctions

rerngvit opened the pull request:
[SPARK-10905][SparkR]: Export freqItems() for DataFrameStatFunctions
- Add function (together with roxygen2 doc) to DataFrame.R and generics.R
- Expose the function in NAMESPACE
- Add unit test for the function

AmplabJenkins commented:
Can one of the admins verify this patch?

shivaram commented:
Jenkins, ok to test

AmplabJenkins commented:
Merged build triggered.

shivaram commented:
cc @sun-rui

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43181 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43181/consoleFull) for   PR 8962 at commit [`f86b6c2`](https://github.com/apache/spark/commit/f86b6c2c5eebb296c9ed53c39e31779e7c45b132).

SparkQA commented:
[Test build #43181 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43181/console) for   PR 8962 at commit [`f86b6c2`](https://github.com/apache/spark/commit/f86b6c2c5eebb296c9ed53c39e31779e7c45b132).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class UnaryMathExpression(val f: Double => Double, name: String)`
  * `case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") `
  * `case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLOOR") `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43181/
Test FAILed.

rerngvit committed:
[SPARK-10905] Fix R code style issues

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43211/consoleFull) for   PR 8962 at commit [`e15645e`](https://github.com/apache/spark/commit/e15645e7cf8c9046619525cd8fd99e8cb97bf65f).

SparkQA commented:
[Test build #43211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43211/console) for   PR 8962 at commit [`e15645e`](https://github.com/apache/spark/commit/e15645e7cf8c9046619525cd8fd99e8cb97bf65f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class HyperLogLogPlusPlus(child: Expression, relativeSD: Double = 0.05)`
  * `case class JsonTuple(children: Seq[Expression])`
  * `abstract class UnaryMathExpression(val f: Double => Double, name: String)`
  * `case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") `
  * `case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLOOR") `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43211/
Test PASSed.

sun-rui commented on R/pkg/R/generics.R:
```diff
@@ -63,6 +63,10 @@ setGeneric("countByValue", function(x) { standardGeneric("countByValue") })
 
+# @rdname freqItems
+# @export
+setGeneric("freqItems", function(x, ..., support = 0.01) { standardGeneric("freqItems") })
```
I think using vector for cols is more R stylish.
setGeneric("freqItems", function(x, cols, support = 0.01) { standardGeneric("freqItems") })

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1873,3 +1872,29 @@ setMethod("as.data.frame",
+#' fi = freqItems(df, "title", "gender")
+#' }
+setMethod("freqItems", signature(x = "DataFrame"),
+          function(x, ..., support) {
```
I think using vector for cols is more R stylish. 
signature(x = "DataFrame", cols = "character", support = "numeric"),
function(x, cols, support = 0.01) {

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1873,3 +1872,29 @@ setMethod("as.data.frame",
+#' Using the frequent element count algorithm described in
+#' \url{http://dx.doi.org/10.1145/762471.762473}, proposed by Karp, Schenker, and Papadimitriou.
+#' 
+#' @param col A list of columns to search frequent items in.
```
Miss the param for the DataFrame.

sun-rui commented on R/pkg/inst/tests/test_sparkSQL.R:
```diff
@@ -1322,6 +1322,19 @@ test_that("crosstab() on a DataFrame", {
+  expect_true("1" %in% multiColResults$letters[[1]])
+  singleColResult <- freqItems(df, "negDoubles", support=0.1)
+  expect_true(-1 %in% head(singleColResult$negDoubles)[[1]])
+})
```
Could you add one more test case, which is R version of the scala test case for freqItems(), like the following:

  l <- lapply(c(0:99), function(i) { if (i %% 2 == 0) { list(1L, -1.0) } else { list(i, i * -1.0) }})
  df <- createDataFrame(sqlContext, l, c("a", "b"))
  result <- collect(freqItems(df, c("a", "b"), 0.4))
  expect_identical(result[[1]], list(list(1L, 99L)))
  expect_identical(result[[2]], list(list(-1, -99)))

sun-rui commented:
@shivaram, could you merge my PR for SPARK-10752 first? that PR creates a new R file for stat functions. Then @rerngvit can rebase this PR to it.

rerngvit commented:
@sun-rui I revised according to your comments. Please have a look.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43225/consoleFull) for   PR 8962 at commit [`57cc150`](https://github.com/apache/spark/commit/57cc150e9cc78c0ac87bd347308e684e5f876f5f).

SparkQA commented:
[Test build #43225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43225/console) for   PR 8962 at commit [`57cc150`](https://github.com/apache/spark/commit/57cc150e9cc78c0ac87bd347308e684e5f876f5f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43225/
Test PASSed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1873,3 +1872,30 @@ setMethod("as.data.frame",
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
```
I think `sqlCtx` should be `sqlContext`

felixcheung replied:
any update?

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1873,3 +1872,30 @@ setMethod("as.data.frame",
+#' fi = freqItems(df, c("title", "gender"))
+#' }
+setMethod("freqItems", signature(x = "DataFrame", cols = "character"),
+          function(x, cols, support) {
```
if support is optional shouldn't it have a default value?

sun-rui replied:
yes, support has a default value. support = 0.01

felixcheung replied:
any update?

sun-rui commented:
@rerngvit, could you rebase it to the latest master? As I moved stats functions to a new R file stats.R in https://github.com/apache/spark/pull/8869

rerngvit committed:
[SPARK-10905] Rebase to master

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43406 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43406/consoleFull) for   PR 8962 at commit [`ae55b24`](https://github.com/apache/spark/commit/ae55b24e062073b36b7b9488fccfe619aa0202a6).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

rerngvit committed:
[SPARK-10905] Revised according to comments
    - Move code to stats.R
    - Revised @rdname document
    - Document the x dataframe
    - Add additional testcase
    - Convert input columns to a vector instead of R ellipsis

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43410/consoleFull) for   PR 8962 at commit [`1339e24`](https://github.com/apache/spark/commit/1339e24e45eae6c09964d4e0808445e176c6feee).

SparkQA commented:
[Test build #43406 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43406/console) for   PR 8962 at commit [`ae55b24`](https://github.com/apache/spark/commit/ae55b24e062073b36b7b9488fccfe619aa0202a6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43406/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43408/
Test FAILed.

SparkQA commented:
[Test build #43410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43410/console) for   PR 8962 at commit [`1339e24`](https://github.com/apache/spark/commit/1339e24e45eae6c09964d4e0808445e176c6feee).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43410/
Test FAILed.

shivaram commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43412/
Test FAILed.

rerngvit commented:
The errors seem not related to the PR. There might be an issue with Jenkins.
"ERROR: Timeout after 15 minutes
ERROR: Error fetching remote repo 'origin'"

shivaram commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43431 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43431/consoleFull) for   PR 8962 at commit [`1339e24`](https://github.com/apache/spark/commit/1339e24e45eae6c09964d4e0808445e176c6feee).

SparkQA commented:
[Test build #43431 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43431/console) for   PR 8962 at commit [`1339e24`](https://github.com/apache/spark/commit/1339e24e45eae6c09964d4e0808445e176c6feee).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43431/
Test PASSed.

sun-rui commented on R/pkg/inst/tests/test_sparkSQL.R:
```diff
@@ -1341,6 +1341,29 @@ test_that("cov() and corr() on a DataFrame", {
+  expect_true(-1 %in% head(singleColResult$negDoubles)[[1]])
+})
+
+test_that("freqItems2() on a DataFrame", {
```
Don't need to add a new test_that. Just move this test case into the above one.

sun-rui commented on R/pkg/R/stats.R:
```diff
@@ -100,3 +100,30 @@ setMethod("corr",
+#' fi = freqItems(df, c("title", "gender"))
+#' }
+setMethod("freqItems", signature(x = "DataFrame", cols = "character"),
+          function(x, cols, support) {
```
support = 0.01

felixcheung replied:
+1

felixcheung commented on R/pkg/R/stats.R:
```diff
@@ -100,3 +100,30 @@ setMethod("corr",
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
```
sqlCtx should be sqlContext

rerngvit committed:
[SPARK-10905] revised according to comments

rerngvit commented:
@sun-rui @felixcheung Thanks for the review. I revised according to your comments. Please have a look.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43463 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43463/consoleFull) for   PR 8962 at commit [`404c8b2`](https://github.com/apache/spark/commit/404c8b2e99397bd62c92cc817cdad4651aec8b42).

SparkQA commented:
[Test build #43463 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43463/console) for   PR 8962 at commit [`404c8b2`](https://github.com/apache/spark/commit/404c8b2e99397bd62c92cc817cdad4651aec8b42).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43463/
Test PASSed.

sun-rui commented:
LGTM

shivaram commented:
Thanks @rerngvit and @sun-rui - LGTM. Merging this
